src/UXClient/Components/HierarchyNavigation/HierarchyNavigation.ts

```diff
@@ -856,12 +855,106 @@ class HierarchyNavigation extends Component{
         return data;
     }
 
-    private closeContextMenu() {
-        if(this.clickedInstance && this.clickedInstance.contextMenu) {
-            this.clickedInstance.contextMenu.remove();
+    public closeContextMenu = () => {
+        if(this.clickedInstance && this.contextMenu) {
+            this.contextMenu.remove();
             d3.selectAll('li.tsi-selected').classed('tsi-selected', false);
         }
         d3.selectAll('.tsi-modelResultWrapper').classed('tsi-selected', false);
+        this.clickedInstance = null;
+    }
+
+    private prepareForContextMenu = (instanceObj, target, wrapperMousePos, eltMousePos) => {
+        let contextMenuProps = {};
+        contextMenuProps['target'] = target;
+        contextMenuProps['wrapperMousePos'] = wrapperMousePos;
+        contextMenuProps['eltMousePos'] = eltMousePos;
+        this.contextMenuProps = contextMenuProps;
+
+        this.clickedInstance = instanceObj;
+        instanceObj.node.classed('tsi-selected', true);
+    }
+
+    public drawContextMenu = (contextMenuItems, contextMenuOptions: ContextMenuOptions) => {
+        let itemList = [];
+        let contextMenuList;
+        this.contextMenu = this.contextMenuProps['target'].append('div').classed('tsi-hierarchyNavigationContextMenu', true).attr('style', () => `top: ${this.contextMenuProps['wrapperMousePos'] - this.contextMenuProps['eltMousePos']}px`);
+        let renderList = (contextMenuItems) => {
+            if (this.contextMenu.select("ul").empty()) {
+                contextMenuList = this.contextMenu.append('ul');
+            } else {
+                this.contextMenu.select("ul").text('');
+            }
+            
+            contextMenuItems.forEach(item => {
```

Comment:
perhaps this should be strongly typed.  one day when we play nicely with typings that will be valuable here.  also, this is a breaking change, but whatev